futures-util/src/compat/compat01as03.rs

```diff
@@ -95,6 +108,83 @@ impl<St: Stream01> Stream03 for Compat01As03<St> {
     }
 }
 
+/// Converts a futures 0.1 Sink object to a futures 0.3-compatible version
+#[derive(Debug)]
+#[must_use = "futures do nothing unless polled"]
+pub struct Compat01As03Sink<S: Sink01> {
+    pub(crate) inner: Spawn01<S>,
+    pub(crate) buffer: Option<S::SinkItem>,
+}
+
+impl<S: Sink01> Unpin for Compat01As03Sink<S> {}
+
+impl<S: Sink01> Compat01As03Sink<S> {
+    /// Wraps a futures 0.1 Sink object in a futures 0.3-compatible wrapper.
+    pub fn new(sink: S) -> Compat01As03Sink<S> {
+        Compat01As03Sink {
+            inner: spawn01(sink),
+            buffer: None,
+        }
+    }
+
+    fn in_notify<R>(&mut self, lw: &LocalWaker, f: impl FnOnce(&mut S) -> R) -> R {
+        let notify = &WakerToHandle(lw.as_waker());
+        self.inner.poll_fn_notify(notify, 0, f)
+    }
+}
+
+impl<S: Sink01> Sink03 for Compat01As03Sink<S> {
+    type SinkItem = S::SinkItem;
+    type SinkError = S::SinkError;
+
+    fn start_send(mut self: Pin<&mut Self>, item: Self::SinkItem) -> Result<(), Self::SinkError> {
+        self.buffer = Some(item);
```

Comment:
Could do with an `assert!(self.buffer.is_none())` before here (maybe `debug_assert!` since it’s a non-`unsafe` programmer error).